eth_p2p/blockchain_sync.nim

```diff
@@ -72,25 +72,32 @@ proc availableWorkItem(ctx: SyncContext): int =
   ctx.workQueue[result] = WantedBlocks(startIndex: nextRequestedBlock, numBlocks: numBlocks.uint, state: Initial)
 
 proc persistWorkItem(ctx: SyncContext, wi: var WantedBlocks) =
-  ctx.chain.persistBlocks(wi.headers, wi.bodies)
+  try:
+    ctx.chain.persistBlocks(wi.headers, wi.bodies)
+    ctx.finalizedBlock = wi.endIndex
+    wi.state = Persisted
+  except:
```

Comment:
avoid blanket catch-alls - know what can go wrong and handle it explicitly. when a random error happens, it's likely that there is a deeper underlying reason that needs attention, and that state is corrupt - this opens up the potential for larger security issues.

Reply:
> know what can go wrong and handle it explicitly

In this case, the exception sources are those doAssert() calls in https://github.com/status-im/nimbus/pull/199

They're all known to be recoverable by discarding the WorkItem and downloading it again.

Reply:
We don't use `doAssert` for recoverable errors. We use it indicate situations that can be considered as bugs in the code - a wrong usage of an API, a detected corruption in the state of the program, etc. 

Recoverable errors are situations that may arise even in a correctly implemented program (e.g. a malicious data coming from the network). The software should have a policy how such situations are handled gracefully in the code. We use error codes or custom exceptions to signal such a condition.

For much more in-depth discussion of the distinction between recoverable and non-recoverable errors, please see here:
https://gist.github.com/zah/d2d729b39d95a1dfedf8183ca35043b3
https://github.com/status-im/nimbus/wiki/Exceptions-and-errors